src/inmanta/server/services/orchestrationservice.py

```diff
@@ -468,117 +395,155 @@ async def delete_version(self, env: data.Environment, version_id: int) -> Apiret
+
+        # hook up all CADs
+        for f, t in cross_agent_dep:
+            res_obj = rid_to_resource[t.resource_str()]
+            res_obj.provides.append(f.resource_id)
+
+        rids = set(rid_to_resource.keys())
+
+        # Sanity checks
+        superfluous_ids = set(resource_sets.keys()) - rids
         if superfluous_ids:
             raise BadRequest(
                 "The following resource ids provided in the resource_sets parameter are not present "
                 f"in the resources list: {', '.join(superfluous_ids)}"
             )
-        requires_ids = set(provides_tree.keys())
-        if not requires_ids.issubset(resource_ids):
+        if not all_requires.issubset(rids):
             raise BadRequest(
                 "The model should have a dependency graph that is closed and no dangling dependencies:"
-                f" {requires_ids-resource_ids}"
+                f" {all_requires - rids}"
             )
-        # hook up all CADs
-        for f, t in cross_agent_dep:
-            res_obj = rv_dict[t.resource_str()]
-            res_obj.provides.append(f.resource_id)
+
+        return rid_to_resource
+
+    def _get_skipped_for_undeployable(
+        self, resources: abc.Sequence[data.Resource], undeployable_ids: abc.Sequence[ResourceIdStr]
+    ) -> abc.Sequence[ResourceIdStr]:
+        """
+        Return the resources that are skipped_for_undeployable given the full set of resources and
+        the resource ids of the resources that are undeployable.
+
+        :param resources: All resources in the model.
+        :param undeployable_ids: The ids of the resource that are undeployable.
+        """
+        # Build up provides tree
+        provides_tree: Dict[ResourceIdStr, List[ResourceIdStr]] = defaultdict(lambda: [])
+        for r in resources:
+            if "requires" in r.attributes:
+                for req in r.attributes["requires"]:
+                    req_id = Id.parse_id(req)
+                    provides_tree[req_id.resource_str()].append(r.resource_id)
+        # Find skipped for undeployables
+        work = list(undeployable_ids)
+        skippeable: Set[ResourceIdStr] = set()
+        while len(work) > 0:
+            current = work.pop()
+            if current in skippeable:
+                continue
+            skippeable.add(current)
+            work.extend(provides_tree[current])
+        return list(skippeable - set(undeployable_ids))
+
+    async def _put_version(
+        self,
+        env: data.Environment,
+        version: int,
+        rid_to_resource: Dict[ResourceIdStr, data.Resource],
+        unknowns: abc.Sequence[data.UnknownParameter],
+        version_info: Optional[JsonType] = None,
+        resource_sets: Optional[Dict[ResourceIdStr, Optional[str]]] = None,
+        partial_base_version: Optional[int] = None,
+        removed_resource_sets: Optional[List[str]] = None,
+        *,
+        connection: asyncpg.connection.Connection,
+    ) -> None:
```

Comment:
Correct

Reply:
One more thing I find hard to reason on at the moment is which invariants are assumed to hold by this method and which are enforced. I think it would be good if the docstring clearly states the extent of this method's responsibilities. Otherwise I think it will be very hard to follow the flow in a few months.

i.e. does it check any of the partial compile invariants or are those are assumed to be met already (the second I think)? How does it handle requires-provides relations / undeployables / purge-on-delete / ... and which invariants of these properties are a precondition?

Reply:
Thanks, I feel like the docstring adds a lot of clarity as to the responsibilities of the different parts.